tests/upgrade_path/test_upgrade_path.py

```diff
@@ -133,17 +145,65 @@ def get_reboot_type(duthost):
         reboot_type = "warm-reboot"
     return reboot_type
 
+
 def check_sonic_version(duthost, target_version):
     current_version = duthost.image_facts()['ansible_facts']['ansible_image_facts']['current']
     assert current_version == target_version, \
         "Upgrade sonic failed: target={} current={}".format(target_version, current_version)
 
 
-def install_sonic(duthost, image_url):
-    res = duthost.reduce_and_add_sonic_images(new_image_url=image_url)
+def install_sonic(duthost, image_url, tbinfo):
+    mg_gwaddr = duthost.get_extended_minigraph_facts(tbinfo).get("minigraph_mgmt_interface", {}).get("gwaddr")
+    mg_gwaddr = ipaddress.IPv4Address(mg_gwaddr)
+    rtinfo_v4 = duthost.get_ip_route_info(ipaddress.ip_network(u'0.0.0.0/0'))
+    for nexthop in rtinfo_v4['nexthops']:
+        if mg_gwaddr == nexthop[0]:
+            new_route_added = False
+            break
+    else:
+        # Temporarily change the default route to mgmt-gateway address. This is done so that
+        # DUT can download an image from a remote host over the mgmt network.
+        logger.info("Add default mgmt-gateway-route to the device via {}".format(mg_gwaddr))
+        duthost.shell("ip route add default via {}".format(mg_gwaddr), module_ignore_errors=True)
+        new_route_added = True
+
+    if urlparse(image_url).scheme not in ('http', 'https',):
+        out = duthost.command("df -BM --output=avail /host",
+                        module_ignore_errors=True)["stdout"]
+        avail = int(out.split('\n')[1][:-1])
+        save_as = "/host/downloaded-sonic-image" if avail >= 2000 else "/tmp/tmpfs/downloaded-sonic-image"
```

Comment:
I am not sure if I understood this correctly. I have a feeling that this logic is: 1. check available space; 2. put something in the directory; 3. check available space again in `reduce_and_add_sonic_images`. The available space in 1 and 3 determines the path of `save_as` which has to be the same in `test_upgrade_path.py` and `reduce_and_add_sonic_images`. However, as the logic probably put something in `/host/downloaded-sonic-image`, the available space might change, and there is possibly a mismatch in `save_as`.

Reply:
In the scenario that a path is given instead of a remote url.

Reply:
This is a valid concern and needs to be fixed. Moreover, I also notice that at this stage `tmpfs` dir is not mounted and hence copy to `/tmp/tmpfs/downloaded-sonic-image` can fail too.

I will update this. The concern is good part of the code is handled in Ansible and Pytest both. I don't know a way in Ansible module to copy remote src file to DUT. If this is possible, I can move this part of the Pytest code (`duthost.copy()`)to `reduce_and_add_sonic_images` module. This will reduce code duplication and add simplicity.